### PR TITLE
feat: build options for zip

### PR DESCRIPTION
**Pull request by Nelsonochoam**
**Description of changes:**
Changes the configuration schema to include options to specify excludes for the zip build system. By default the zip build system grabs all the files from the root folder (excluding a few) when the excludes option is specified we will only exclude greengrass-build, recipe and gdk-config files + whatever the customer configures.

**Why is this change necessary:**
Part of functionality being requested by customers

### CODE REVIEW

File: Makefile [hunk 1]
```diff
@@ -0,0 +1,5 @@
+lint:
```
Comment (Nelsonochoam) [inline Makefile:1]: For convenience during local development

File: gdk/static/config_schema.json [hunk 2]
```diff
@@ -64,21 +64,58 @@
                                     "build_system"
                                 ]
                             },
-                            "if": {
-                                "properties": {
-                                    "build_system": {
-                                        "const": "custom"
+                            "allOf": [
```
Comment (Nelsonochoam) [inline gdk/static/config_schema.json:67]: This change allow us to have multi conditional if statements which is now required to accept different options per build system type

https://json-schema.org/understanding-json-schema/reference/conditionals.html

File: tests/gdk/build_system/test_zip_build_configuraiton.py [hunk 3]
```diff
@@ -0,0 +1,60 @@
...
+
+        with pytest.raises(Exception):
+            options = {"excludes": []}
+            validate_configuration(self.configuration_base(options))
+
+        with pytest.raises(Exception):
+            options = {}
+            validate_configuration(self.configuration_base(options))
+
+        with pytest.raises(Exception):
```
Comment (dgggIoT) [inline tests/gdk/build_system/test_zip_build_configuraiton.py:58]: what is the purpose of this sample-like dict?

File: gdk/static/config_schema.json [hunk 4]
```diff
@@ -64,21 +64,58 @@
...
                                 },
-                                "required": [
-                                    "custom_build_command"
-                                ]
-                            }
+                                {
+                                    "if": {
+                                        "properties": {
+                                            "build_system": {
+                                                "const": "custom"
```
Comment (dgggIoT) [inline gdk/static/config_schema.json:105]: what build_system will be considered as "custom"? Will "zip" be considerred as "custom" too?

File: tests/gdk/build_system/test_zip_build_configuraiton.py (hunk 3 above)
Comment (Nelsonochoam) [reply tests/gdk/build_system/test_zip_build_configuraiton.py:58]: That random unsupported value raises an exception

**Review by dgggIoT — DISMISSED** [review]

File: tests/gdk/build_system/test_zip_build_configuraiton.py [hunk 5]
```diff
@@ -0,0 +1,67 @@
...
+                    "build": {
+                        "build_system": "zip"
+                    },
+                    "publish": {
+                        "bucket": "default",
+                        "region": "us-east-1"
+                    },
+                }
+            },
+            "gdk_version": "1.0.0",
```
Comment (saranyailla) [inline tests/gdk/build_system/test_zip_build_configuraiton.py:28]: Nit: Change this to 1.2.0 though it doesn't really matter.

File: tests/gdk/build_system/test_zip_build_configuraiton.py [hunk 6]
```diff
@@ -0,0 +1,67 @@
...
+        }
+
+        if options is not None:
+            base["component"]["com.build.zip"]["build"].setdefault("options", options)
+
+        return base
+
+    def test_valid_configuration_options(self):
```
Comment (saranyailla) [inline tests/gdk/build_system/test_zip_build_configuraiton.py:36]: Nit: You can use `pytest.mark.parametrize` for passing multiple inputs in these two tests.

File: gdk/static/config_schema.json [hunk 7]
```diff
@@ -64,21 +64,50 @@
...
+                                                "const": "zip"
+                                            }
+                                        }
+                                    },
+                                    "then": {
+                                        "properties": {
+                                            "options": {
+                                                "type": "object",
+                                                "description": "configuraiton options for the zip build system",
+                                                "required": ["exclude"],
```
Comment (saranyailla) [inline gdk/static/config_schema.json:81]: I'm not sure how the "options" for "zip" build system are gonna be extended yet. So, I don't think we can go with "required" here for exclude.

Comment (Nelsonochoam) [reply gdk/static/config_schema.json:81]: Right now it is required given it is the only attribute. Moving forward might not be. This config might change a bit in order for us to support zipping multiple folders

Comment (Nelsonochoam) [reply gdk/static/config_schema.json:81]: Making a field required to not required is safe (2 way door) if we want to change it later on. The other way around is not.

**Review by saranyailla — APPROVED** [review]

File: tests/gdk/build_system/test_zip_build_configuraiton.py [hunk 8]
```diff
@@ -0,0 +1,61 @@
+from pathlib import Path
```
Comment (avipinku) [inline tests/gdk/build_system/test_zip_build_configuraiton.py:1]: should be `test_zip_build_configuration.py`

File: gdk/static/config_schema.json [hunk 9]
```diff
@@ -64,21 +64,50 @@
...
+                                            "build_system": {
+                                                "const": "zip"
+                                            }
+                                        }
+                                    },
+                                    "then": {
+                                        "properties": {
+                                            "options": {
+                                                "type": "object",
+                                                "description": "configuraiton options for the zip build system",
```
Comment (avipinku) [inline gdk/static/config_schema.json:80]: nit: `configuration`

**Review by avipinku — APPROVED** [review]

**Review by saranyailla — APPROVED** [review]

**Pull request merged**